compiler/noirc_frontend/src/hir/comptime/interpreter.rs

```diff
@@ -924,310 +932,186 @@ impl<'local, 'interner> Interpreter<'local, 'interner> {
             InterpreterError::InvalidValuesForBinary { lhs, rhs, location, operator }
         };
 
-        use InterpreterError::InvalidValuesForBinary;
-        match infix.operator.kind {
-            BinaryOpKind::Add => match (lhs_value.clone(), rhs_value.clone()) {
-                (Value::Field(lhs), Value::Field(rhs)) => Ok(Value::Field(lhs + rhs)),
-                (Value::I8(lhs), Value::I8(rhs)) => {
-                    Ok(Value::I8(lhs.checked_add(rhs).ok_or(error("+"))?))
-                }
-                (Value::I16(lhs), Value::I16(rhs)) => {
-                    Ok(Value::I16(lhs.checked_add(rhs).ok_or(error("+"))?))
-                }
-                (Value::I32(lhs), Value::I32(rhs)) => {
-                    Ok(Value::I32(lhs.checked_add(rhs).ok_or(error("+"))?))
+        /// Generate matches that can promote the type of one side to the other if they are compatible.
+        ///
+        /// ```ignore
+        /// BinaryOpKind::Subtract => match (lhs_value, rhs_value) {
+        ///     (Value::I16(lhs), Value::I8(rhs)) => {
+        ///         let rhs = rhs as i16;
+        ///         Ok(Value::I16(lhs.checked_sub(rhs).ok_or(error("-"))?))
+        ///     }
+        /// }
+        /// ```
+        macro_rules! match_values {
+            (($lhs_value:ident as $lhs:ident $op:literal $rhs_value:ident as $rhs:ident) {
+                $(
+                    $variant:ident($typ:ty) from [$($promoted:ident),+] to $result:ident => $expr:expr
+                ),*
+                $(,)?
+             }
+            ) => {
```

Comment:
Ah, okay, then I can remove most of these. It was indeed the bit shifting [in sha256](https://github.com/noir-lang/noir/blob/6f47cf14ef4f4d501980ad30f2db711e919ccd26/noir_stdlib/src/hash/sha256.nr#L420-L422) that the interpreter couldn't handle. I didn't check how Noir handles the other cases :-/  

Then I suppose I can narrow it down to only handle the case of shifting. Does Noir always restrict the `rhs` of bit shifting to `u8`?

Reply:
>Then I suppose I can narrow it down to only handle the case of shifting. Does Noir always restrict the rhs of bit shifting to u8?

yep that's right.

Reply:
Changed the macros to only allow different types in case of bit shifting, and only `U8` on the RHS, like Noir. Kept the helpers so the formulas are only stated once (so we have 5 blocks of match arms, rather than 16).